db/dataset.py

```diff
@@ -192,3 +193,116 @@ def delete(id):
     """Delete dataset with a specified ID."""
     with db.engine.begin() as connection:
         connection.execute("DELETE FROM dataset WHERE id = %s", (str(id),))
+
+
+def create_snapshot(dataset_id):
+    """Creates a snapshot of current version of a dataset.
+
+    Snapshots are stored as JSON and have the following structure:
+    {
+        "name": "..",
+        "description": "..",
+        "classes": [
+            {
+                "name": "..",
+                "description: "..",
+                "recordings": ["..", ...]
+            },
+            ...
+        ]
+    }
+
+    Args:
+        dataset_id (string/uuid): ID of a dataset.
+
+    Returns:
+        ID (UUID) of a snapshot that was created.
+    """
+    dataset = get(dataset_id)
+    if not dataset:
+        raise exceptions.NoDataFoundException("Can't find dataset with a specified ID.")
+    snapshot = {
+        "name": dataset["name"],
+        "description": dataset["description"],
+        "classes": dataset["classes"],
+    }
+    with db.engine.connect() as connection:
+        result = connection.execute(sqlalchemy.text("""
+            INSERT INTO dataset_snapshot (id, dataset_id, data)
+                 VALUES (uuid_generate_v4(), :dataset_id, :data)
+              RETURNING id
+        """), {
+            "dataset_id": dataset_id,
+            "data": json.dumps(snapshot),
+        })
+        return result.fetchone()["id"]
+
+
+def get_snapshot(id):
+    """Get snapshot of a dataset.
+
+    Args:
+        id (string/uuid): ID of a snapshot.
+
+    Returns:
+        dictionary: {
+            "id": <ID of the snapshot>,
+            "dataset_id": <ID of the dataset that this snapshot is associated with>,
+            "created": <creation time>,
+            "data": <actual content of a snapshot (see `create_snapshot` function)>
+        }
+    """
+    with db.engine.connect() as connection:
+        result = connection.execute(sqlalchemy.text("""
+            SELECT id::text
+                 , dataset_id::text
+                 , data
+                 , created
+              FROM dataset_snapshot
+             WHERE id = :id
+        """), {"id": id})
+        row = result.fetchone()
+        if not row:
+            raise db.exceptions.NoDataFoundException("Can't find dataset snapshot with a specified ID.")
+        return dict(row)
+
+def delete_snapshot(connection, snapshot_id):
```

Comment:
I'm not sure what to do here. Our habit has been to have standalone methods which work in a transaction, but snapshot deletion should happen at the same time as dataset_eval_job deletion, therefore in the same transaction. We could copy the sql code into that method, but then that potentially leaves duplicate code, which is difficult to maintain.
I'm not sure what we should do here. Some options:
 * We currently only delete snapshots when we delete a pending eval job, so just move the sql there now and deal with this problem later
 * Come up with a style guide for these helper methods that can be called with a connection as an argument. Maybe `_delete_snapshot_c(connection, snapshot_id)` and then if we need the method publicly accessible we create `delete_snapshot` which creates a connection and calls this method.